lib/nx_gzlib.c

```diff
@@ -155,6 +156,44 @@ gzFile nx_gzopen_(const char* path, int fd, const char *mode)
 	return file;
 }
 
+int nx_gzread(gzFile file, void *buf, unsigned len)
+{
+	gz_statep state = (gz_statep) file;
+	z_streamp stream = &(state->strm);
+	int ret, err;
+	const uInt max = len > 10? 10 : 1;
+	uLong last_total_out = stream->total_out;
+	unsigned char* next_in = malloc(sizeof(char)*max);
+
+	stream->next_out = buf;
+	stream->avail_out = len;
+	do {
+		ret = read(state->fd, next_in, sizeof(char)*max);
+		stream->next_in = next_in;
+		if (ret == 0){
+			/* Flush all at the end of the file. */
+			err = nx_inflate(stream, Z_FINISH);
+			break;
+		}
+
+		stream->avail_in = ret;
+		err = nx_inflate(stream, Z_NO_FLUSH);
+
+		if (err != Z_OK && err != Z_STREAM_END){
+			state->err = err;
+			return 0;
+		}
+	} while (stream->avail_out != 0);
+
+	if (stream->avail_in > 0){
+		nx_streamp s = (nx_streamp) stream->state;
+		copy_data_to_fifo_in(s);
+	}
```

Comment:
Because `gzread` returns when `len` bytes are wrote on `buf`, if we still have unprocessed data on `avail_in` we need to save it or we will lose it, IMHO `fifo_in` seems the best place to save it. Another option is to skip `read` when `avail_in` is not 0, but we can't guarantee `avail_in` will not be overwritten somewhere else between `gzread` calls.

Reply:
@rzinsly another option is to `lseek` the file offset `avail_in` bytes back, so the next `read` will get the unprocessed bytes. The downside is that we would be re-reading these bytes from disk in the future, while a copy in memory would most likely be cheaper.

Reply:
I like what @rzinsly did here, but I think the code deserves a source code comment that is a copy if his explanation here.

Reply:
I added a source code comment, `lseek` is not an option because `gzread` needs to be able to read from `stdin`.